From 045468cb9bff47bb3bb72268b6d5a3fe44e383db Mon Sep 17 00:00:00 2001
From: Andrei Cioara <andrei@cioara.me>
Date: Sat, 6 Jul 2019 09:09:47 -0700
Subject: [PATCH] Fixed shell injection issue (#102)

* Providing credentials as env variables now

* Logging the command to the logging for debugging as well

* Removed shell injection from ls

* Removed shell injection possibility from copy as well
---
 src/backend/api/utils/rclone_connection.py | 90 ++++++++++++----------
 1 file changed, 48 insertions(+), 42 deletions(-)

diff --git a/src/backend/api/utils/rclone_connection.py b/src/backend/api/utils/rclone_connection.py
index e194a25f..9afa909c 100644
--- a/src/backend/api/utils/rclone_connection.py
+++ b/src/backend/api/utils/rclone_connection.py
@@ -6,6 +6,7 @@
 import subprocess
 import threading
 import time
+import os
 
 class RcloneConnection:
     def __init__(self):
@@ -22,10 +23,14 @@ def __init__(self):
 
     def verify(self, data):
         credentials = self._formatCredentials(data, name='current')
-        command = '{} rclone lsjson current:'.format(credentials)
+        command = [
+            'rclone',
+            'lsjson',
+            'current:',
+        ]
 
         try:
-            result = self._execute(command)
+            result = self._execute(command, credentials)
             return {
                 'result': True,
                 'message': 'Success',
@@ -40,17 +45,14 @@ def verify(self, data):
 
     def ls(self, data, path):
         credentials = self._formatCredentials(data, name='current')
-
-        command = (
-            '{credentials} '
-            'rclone lsjson current:{path}'
-        ).format(
-            credentials=credentials,
-            path=path,
-        )
+        command = [
+            'rclone',
+            'lsjson',
+            'current:{}'.format(path),
+        ]
 
         try:
-            result = self._execute(command)
+            result = self._execute(command, credentials)
             result = json.loads(result)
             return result
         except subprocess.CalledProcessError as e:
@@ -61,17 +63,14 @@ def ls(self, data, path):
 
     def mkdir(self, data, path):
         credentials = self._formatCredentials(data, name='current')
-
-        command = (
-            '{credentials} '
-            'rclone touch current:{path}/.keep'
-        ).format(
-            credentials=credentials,
-            path=path,
-        )
+        command = [
+            'rclone',
+            'touch',
+            'current:{}/.keep'.format(path),
+        ]
 
         try:
-            result = self._execute(command)
+            result = self._execute(command, credentials)
             return {
                 'message': 'Success',
             }
@@ -81,33 +80,35 @@ def mkdir(self, data, path):
 
 
     def copy(self, src_data, src_path, dst_data, dst_path, job_id=None):
-        credentials = ''
+        credentials = {}
 
         if src_data is None: # Local
             src = src_path
         else:
-            credentials += self._formatCredentials(src_data, name='src')
+            credentials.update(self._formatCredentials(src_data, name='src'))
             src = 'src:{}'.format(src_path)
 
         if dst_data is None: # Local
             dst = dst_path
         else:
-            credentials += self._formatCredentials(dst_data, name='dst')
+            credentials.update(self._formatCredentials(dst_data, name='dst'))
             dst = 'dst:{}'.format(dst_path)
 
+        command = [
+            'rclone',
+            'copy',
+            src,
+            dst,
+            '--progress',
+            '--stats', '2s',
+        ]
 
-        command = (
-            '{credentials} '
-            'rclone copy {src} {dst} '
-            '--progress '
-            '--stats 2s '
-        ).format(
-            credentials=credentials,
-            src=src,
-            dst=dst,
+        bash_command = "{} {}".format(
+            ' '.join("{}='{}'".format(key, value) for key, value in credentials.items()),
+            ' '.join(command),
         )
 
-        logging.info(sanitize(command))
+        logging.info(sanitize(bash_command))
 
         if job_id is None:
             job_id = self._get_next_job_id()
@@ -118,7 +119,7 @@ def copy(self, src_data, src_path, dst_data, dst_path, job_id=None):
         self._stop_events[job_id] = threading.Event()
 
         try:
-            self._execute_interactive(command, job_id)
+            self._execute_interactive(command, credentials, job_id)
         except subprocess.CalledProcessError as e:
             raise RcloneException(sanitize(str(e)))
 
@@ -154,13 +155,13 @@ def _formatCredentials(self, data, name):
 
         prefix = "RCLONE_CONFIG_{}".format(name.upper())
 
-        credentials = ''
-        credentials += "{}_TYPE='{}' ".format(prefix, data.type)
+        credentials = {}
+        credentials['{}_TYPE'.format(prefix)] = data.type
 
         def _addCredential(credentials, env_key, data_key):
             value = getattr(data, data_key, None)
             if value is not None:
-                credentials += "{}='{}' ".format(env_key, value)
+                credentials[env_key] = value
             return credentials
 
 
@@ -253,30 +254,35 @@ def _job_id_exists(self, job_id):
         return job_id in self._job_status
 
 
-    def _execute(self, command):
-        byteOutput = subprocess.check_output(command, shell=True)
+    def _execute(self, command, env={}):
+        full_env = os.environ.copy()
+        full_env.update(env)
+        byteOutput = subprocess.check_output(command, env=full_env)
         output = byteOutput.decode('UTF-8').rstrip()
         return output
 
 
-    def _execute_interactive(self, command, job_id):
+    def _execute_interactive(self, command, env, job_id):
         thread = threading.Thread(target=self.__execute_interactive, kwargs={
             'command': command,
+            'env': env,
             'job_id': job_id,
         })
         thread.daemon = True
         thread.start()
 
 
-    def __execute_interactive(self, command, job_id):
+    def __execute_interactive(self, command, env={}, job_id=0):
         stop_event = self._stop_events[job_id]
+        full_env = os.environ.copy()
+        full_env.update(env)
 
         process = subprocess.Popen(
             command,
+            env=full_env,
             stdin=subprocess.PIPE,
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,
-            shell=True,
         )
 
         reset_sequence1 = '\x1b[2K\x1b[0' # + 'G'
